src/FrequencyBalancerAgent.cpp

```diff
@@ -0,0 +1,644 @@
+        , m_power_gov(power_gov)
+        , m_freq_governor(frequency_gov)
+        , m_sst_clos_governor(sst_gov)
+        , m_frequency_ctl_domain_type(m_freq_governor->frequency_domain_type())
+        , m_frequency_control_domain_count(m_platform_topo.num_domain(m_frequency_ctl_domain_type))
+        , m_network_hint_sample_length(m_frequency_control_domain_count, 0)
+        , m_non_network_hint_sample_length(m_frequency_control_domain_count, 0)
+        , m_last_hp_count(2, 0)
+        , m_handle_new_epoch(false)
+        , m_epoch_wait_count(MINIMUM_EPOCHS_FOR_NEW_EPOCH_CONTROL)
+        , m_package_balancers(package_balancers)
+        , m_frequency_limit_detector(frequency_limit_detector)
+    {
+        if (m_package_balancers.empty()) {
+            for (int i = 0; i < m_package_count; ++i) {
+                // We balance each CPU package independently so we can take
+                // advantage of power headroom within each package. Create
+                // one balancer per package.
+                m_package_balancers.push_back(FrequencyTimeBalancer::make_unique(
+                    m_frequency_min, m_frequency_max, m_frequency_step));
+            }
+        }
+        for (int i = 0; i < m_package_count; ++i) {
+            // Determine which control indices (e.g., CPUs or cores) map to
+            // each balancer. We need this since some operations (e.g., platform IO)
+            // are made with respect to all controls in a domain, and others (e.g.,
+            // intra-package rebalancing) operate on subsets.
+            auto control_indices_in_package = m_platform_topo.domain_nested(
+                m_frequency_ctl_domain_type, GEOPM_DOMAIN_PACKAGE, i);
+            m_package_core_indices.emplace_back(
+                control_indices_in_package.begin(), control_indices_in_package.end());
+        }
+        if (!m_frequency_limit_detector) {
+            m_frequency_limit_detector = FrequencyLimitDetector::make_unique(
+                m_platform_io, m_platform_topo);
+        }
+
+        if (m_sst_clos_governor) {
+            m_frequency_ctl_domain_type = m_sst_clos_governor->clos_domain_type();
+            m_frequency_control_domain_count = m_platform_topo.num_domain(m_frequency_ctl_domain_type);
+            m_network_hint_sample_length.resize(m_frequency_control_domain_count, 0);
+            m_non_network_hint_sample_length.resize(m_frequency_control_domain_count, 0);
+        }
+    }
+
+    std::string FrequencyBalancerAgent::plugin_name(void)
+    {
+        return "frequency_balancer";
+    }
+
+    std::unique_ptr<Agent> FrequencyBalancerAgent::make_plugin(void)
+    {
+        return geopm::make_unique<FrequencyBalancerAgent>();
+    }
+
+    void FrequencyBalancerAgent::init(int level, const std::vector<int> &fan_in,
+                                     bool is_level_root)
+    {
+        static_cast<void>(is_level_root);
```

Comment:
? not sure what's going on here.  is_level_root isn't stored after the cast or used.

Reply:
At one point, one of the compilers I was using on this project emitted warnings about unused variables. The void-cast signals to the compiler not to warn since this is an intentionally unused field. This agent doesn't do anything special based on whether the agent instance acts as the root instance in the current level of the tree.